2.0.0

mindplay-dk opened the pull request.

mindplay-dk committed:
test fails due to #4

mindplay-dk committed:
component are now consistently immutable once activated, per #4
BREAKING CHANGE for 2.0

mindplay-dk committed:
Merge branch 'enforce-integrity' into feature/cloneable

mindplay-dk committed:
add support for cloning
simplify immutability model

mindplay-dk committed:
internal optimization: avoid wrapping simple values in closures

mindplay-dk committed:
initialize is an implementation detail (it can't safely be called by a superclass)

mindplay-dk committed:
tweak

mindplay-dk committed:
avoid unsafe constructor injections where parameter names happen to match registered component names, per #5
BREAKING CHANGE: although no one seemed to like this undocumented feature - it's unlikely you were using it, but if you were,
you need to adjust those registrations to explicitly map constructor parameter names/types using calls to ref()

mindplay-dk committed:
add 2.x upgrade guide

mindplay-dk committed:
tweak

mindplay-dk committed:
preliminary refactoring: extract common reflection concerns to a pseudo-namespace

mindplay-dk committed:
separation of bootstrapping/configuration into a factory layer - this provides a naturally immutable container (and reusable bootstrapping)

mindplay-dk committed:
remove duplicate test

mindplay-dk committed:
add docs

mindplay-dk committed:
remove state-dependent name-inference from configure()

mindplay-dk committed:
eliminate has() from ContainerFactory (as using it implies creation of factory state conditions, which shouldn't exist)

mindplay-dk committed:
adding an empty constructor (since there's no default constructor in PHP, and we don't want an eventual addition of a constructor to create a BC break)

mindplay-dk committed:
add internal Container::inject() method

mindplay-dk committed:
extract Configuration base-class to eliminate duplication and simplify initialization
(minor BC break from previous 2.0-branch commit)

rhl-jfm commented on composer.json:
```diff
@@ -15,7 +15,7 @@
     "require-dev": {
         "mindplay/benchpress": "dev-master",
-        "mindplay/testies": "^0.2",
+        "mindplay/testies": "dev-master",
```
Do you really need to depend on dev-master? Wouldn't it be nicer to release a new version of testies perhaps?

mindplay-dk replied:
Yeah, that's been at the bottom of my infinite todo-list for a few years ;-)

mindplay-dk committed:
PHP 7.1.0RC2 has a reflection issue - changing condition for the test to PHP 7.1.0RC3
removing PHP 7.1 build target from Travis for now, as they're still running RC2
